docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,375 @@
+    pub common: Attributes,
+    pub header: Option<AnyHeader>,
+}
+```
+
+This can be solved using the `ibc_proto::google::protobuf::Any` type instead and having the light client traits provide
+an encoding to `Any` ->
+
+```rust
+use ibc_proto::google::protobuf::Any;
+
+pub trait Header {
+    /// Encode to canonical protobuf representation
+    fn encode_any(&self) -> Any;
+
+    /* ... */
+}
+```
+
+Thankfully, the core modules code doesn't use the `serde` derivations except in logs and errors. Host and light-client
+implementations can optionally choose to downcast to the concrete type and use it's `serde` derivations directly (if
+available).
+
+#### Light client traits cannot have constructors
+
+This restriction comes from the fact that trait methods of object safe traits cannot return `Self`.  
+However, we would need a constructor to be able to create a `ClientState` and `ConsensusState` in the `create_client`
+handler. This can be done using a `where Self: Sized` clause on the trait method.
+
+```rust
+use ibc_proto::google::protobuf::Any;
+
+pub trait ClientState {
+    fn decode(any: Any) -> Box<dyn ClientState> where Self: Sized;
+
+    /* ... */
+}
+```
+
+### Downcasting support
+
+We need the ability to upcast `&dyn T` to `&dyn core::any::Any` to be able to downcast to a concrete type. This can be
+done using a trait with a blanket implementation ->
+
+```rust
+pub mod dynamic_typing {
+    use core::any::Any;
+
+    pub trait AsAny: Any {
+        fn as_any(&self) -> &dyn Any;
+    }
+
+    impl<T: Any> AsAny for T {
+        fn as_any(&self) -> &dyn Any {
+            self
+        }
+    }
+}
+```
+
+All light client traits could then add `AsAny` as a supertrait.
+
+```rust
+use crate::dynamic_typing::AsAny;
+
+pub trait Header: AsAny { /* ... */ }
+```
+
+And usage would look like ->
+
+```rust
+fn downcast_header(h: &dyn Header) -> Result<&TmHeader, Ics02Error> {
+    h.as_any()
+        .downcast_ref::<TmHeader>()
+        .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))
+}
+```
+
+### Special case: removing `AnyClient`
+
+The `AnyClient` enum is special as it is mostly stateless and always created on-the-fly (using `ClientType`) during
+verification in the handlers. e.g. `update_client::process()` ->
+
+```rust
+pub fn process(
+    ctx: &dyn ClientReader,
+    /* ... */
+) -> HandlerResult<ClientResult, Error> {
+    /* ... */
+
+    let client_type = ctx.client_type(&client_id)?;
+
+    let client_def = AnyClient::from_client_type(client_type);
+
+    let client_state = ctx.client_state(&client_id)?;
+
+    /* ... */
+}
+```
+
+This is problematic because it means the module code must be aware of a `ClientType` -> `ClientDef` mapping. This can be
+solved by requiring the `ClientState` trait to provide us with its `ClientDef` implementation.
+
+```rust
+pub trait ClientState {
+    /// Return the associated `ClientDef` implementation
+    fn client_def(&self) -> Box<dyn ClientDef>;
```

Comment:
Instead of having `ClientState` provide the `client_def()`, why not have `ClientReader` have a `client_def(client_id)` method instead?

As discussed, this is a different way of thinking about it. I see the `ClientState` as simply encapsulating the parameters of a light client. To me, it's more natural for the chain/context to know the type of all its clients (indexed by `clientId`), rather than the parameters knowing about their corresponding client implementation.

Reply:
I see your point, I have been thinking about this myself as well. To me, the `ClientState` represents an instance of a light client and it's the only 1:1 mapping of a client id v/s light client that is stored on-chain. ~~And the `ClientState` trait is something that the light client will implement. So it makes more sense to me to ask the light client to give us access to its verification functions instead of asking the host. Besides, if `client_def()` was a method on `ClientReader`, that would require the host to maintain a `ClientId` <> `ClientType` mapping.~~
Just checked the spec and it already requires hosts to store a `ClientId` <> `ClientType` mapping and your interpretation seems to be more in line with the spec.

Reply:
d66271656205207c92e497a5c69f0708bbba469e